sdk/communication/azure-communication-chat/src/main/java/com/azure/android/communication/chat/implementation/notifications/fcm/RegistrationKeyStore.java

```diff
@@ -0,0 +1,154 @@
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.Reader;
+import java.lang.reflect.Type;
+import java.util.ArrayList;
+import java.util.Deque;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Stack;
+
+import javax.crypto.SecretKey;
+import javax.crypto.spec.SecretKeySpec;
+
+/**
+ * This class represents a storage facility for secrete key as string and its creation time. The
+ * data is persisted in a file as json format. The key-store itself could load data from file for
+ * faster operation.
+ *
+ * The file persists all the data in Json format. Latest record is added to last of the list. For example:
+ * [
+ *     {
+ *         "abcdefgh",
+ *         "ijklmno"
+ *         "1234";
+ *     },
+ *     {
+ *         "oklliabe",
+ *         "abcdefgh",
+ *         "1236";
+ *     }
+ * ]
+ */
+public class RegistrationKeyStore {
+    private List<RegistrationKeyEntry> list;
+
+    private ClientLogger clientLogger = new ClientLogger(RegistrationKeyStore.class);
+
+    public RegistrationKeyStore() {
+        list = new ArrayList<>();
+    }
+
+    public int getSize() {
+        return list.size();
+    }
+
+    //Loading data with given path to memory
+    public void load(InputStream inputStream) {
+        //Nothing to read from
+        if (inputStream == null) {
+            return;
+        }
+        try {
+            Gson gson = new Gson();
+            Type type = new TypeToken<List<RegistrationKeyEntry>>() {}.getType();
+            Reader reader = new InputStreamReader(inputStream);
+            list = gson.fromJson(reader, type);
+            reader.close();
+        } catch (IOException e) {
+            clientLogger.logExceptionAsError(new RuntimeException(e));
+        }
+    }
+
+    //Write in-memory map into file with Json format
+    private void writeJsonToFile(String path) {
+        File outputFile = new File(path);
+        try {
+            outputFile.createNewFile();
+        } catch (IOException e) {
+            clientLogger.logExceptionAsError(new RuntimeException("Failed to create key store file", e));
+        }
+        Gson gson = new Gson();
+        String jsonStr = gson.toJson(list, list.getClass());
+
+        try(FileOutputStream fos = new FileOutputStream(outputFile);
+            BufferedOutputStream bos = new BufferedOutputStream(fos)) {
+            //convert string to byte array
+            byte[] bytes = jsonStr.getBytes();
+            //write byte array to file
+            bos.write(bytes);
+            bos.close();
+            fos.close();
+        } catch (IOException e) {
+            clientLogger.logExceptionAsError(new RuntimeException("Filed writing key map to file", e));
+        }
+    }
+
+    public void storeKeyEntry(String path, RegistrationKeyEntry registrationKeyEntry) {
+        list.add(list.size(), registrationKeyEntry);
+        writeJsonToFile(path);
```

Comment:
How often do we store a new key entry?

Reply:
We renew the registration every 15 minutes. And will add 1-3 entries based on retry attempts. So the TTL is set as 45 minutes (2700 seconds). Registra team is going to change the TTL on their to around 50 minutes.